### ISSUE TITLE
How to load espnet2 models in Pytorch

### ISSUE THREAD

**Issue by sun-peach** [espnet/espnet#1939]
Hi, I have ran the egs2/aishell recipe and get the end2end ASR model.

Now I would like to use the trained model as a sub-model in my big neural network. Its output will be fed to other layers for further processing. I tried to load the model, however, I found that, in the `class ESPnetASRModel(AbsESPnetModel)`, the forward function is outputing loss rather than the expected text embedding (or say the score vectors).

 ```
   def forward(
        self,
        speech: torch.Tensor,
        speech_lengths: torch.Tensor,
        text: torch.Tensor,
        text_lengths: torch.Tensor,
    ) -> Tuple[torch.Tensor, Dict[str, torch.Tensor], torch.Tensor]:
        """Frontend + Encoder + Decoder + Calc loss
        Args:
            speech: (Batch, Length, ...)
            speech_lengths: (Batch, )
            text: (Batch, Length)
            text_lengths: (Batch,)
        """
        assert text_lengths.dim() == 1, text_lengths.shape
        # Check that batch_size is unified
        assert (
            speech.shape[0]
            == speech_lengths.shape[0]
            == text.shape[0]
            == text_lengths.shape[0]
        ), (speech.shape, speech_lengths.shape, text.shape, text_lengths.shape)
        batch_size = speech.shape[0]
        # for data-parallel
        text = text[:, : text_lengths.max()]
        # 1. Encoder
        encoder_out, encoder_out_lens = self.encode(speech, speech_lengths)
        # 2a. Attention-decoder branch
        if self.ctc_weight == 1.0:
            loss_att, acc_att, cer_att, wer_att = None, None, None, None
        else:
            loss_att, acc_att, cer_att, wer_att = self._calc_att_loss(
                encoder_out, encoder_out_lens, text, text_lengths
            )
        # 2b. CTC branch
        if self.ctc_weight == 0.0:
            loss_ctc, cer_ctc = None, None
        else:
            loss_ctc, cer_ctc = self._calc_ctc_loss(
                encoder_out, encoder_out_lens, text, text_lengths
            )

        # 2c. RNN-T branch
        if self.rnnt_decoder is not None:
            _ = self._calc_rnnt_loss(encoder_out, encoder_out_lens, text, text_lengths)

        if self.ctc_weight == 0.0:
            loss = loss_att
        elif self.ctc_weight == 1.0:
            loss = loss_ctc
        else:
            loss = self.ctc_weight * loss_ctc + (1 - self.ctc_weight) * loss_att

        stats = dict(
            loss=loss.detach(),
            loss_att=loss_att.detach() if loss_att is not None else None,
            loss_ctc=loss_ctc.detach() if loss_ctc is not None else None,
            acc=acc_att,
            cer=cer_att,
            wer=wer_att,
            cer_ctc=cer_ctc,
        )

        # force_gatherable: to-device and to-tensor if scalar for DataParallel
        loss, stats, weight = force_gatherable((loss, stats, batch_size), loss.device)
        return loss, stats, weight
```

Can anyone tell me how to get the output from the decoder so that I can implement my model? Thank you.

**Comment by kamo-naoyuki** [comment]
You need to define it by yourself, but it's not hard. For example.

```python
def get_output(
    model: ESPnetASRModel,
    speech: torch.Tensor,
    speech_lengths: torch.Tensor,
    text: torch.Tensor,
    text_lengths: torch.Tensor,
) 
    encoder_out, encoder_out_lens = model.encode(speech, speech_lengths)
    ys_in_pad, ys_out_pad = add_sos_eos(text, model.sos, model.eos, model.ignore_id)
    ys_in_lens = text_lengths + 1
    decoder_out, _ = model.decoder(encoder_out, encoder_out_lens, ys_in_pad, ys_in_lens)
    return decoder_out
```

(You can refer the following part too)

https://github.com/espnet/espnet/blob/f36bbee58afe8c718dc465ceeb85b86041801e0c/espnet2/asr/espnet_model.py#L230-L236

**Comment by sun-peach** [comment]
@kamo-naoyuki Thank you very much! It points me a direction. However, when I dig into the details, I still have some questions:
1. the forward function calls the `_calc_att_loss` function with the "text" arguments. I could not find what is the text. Isn't the text supposed to be output from the decoder? Why do we need input it here? (Sorry for noob question) 
https://github.com/espnet/espnet/blob/f36bbee58afe8c718dc465ceeb85b86041801e0c/espnet2/asr/espnet_model.py#L121-L124
2. Is there anyway that I do not have the text output, but the output from the layer before the decoder output layer (I only use it for representation extraction)?

Thank you very much!

**Comment by kamo-naoyuki** [comment]
`text` is a sequence of integers representing the token IDs of the transcription.

Please explain your situation more.
What is representation extraction? Embedding vector from token ID?
Or, do you mean, you need  the output vector from decoder when recognition?

**Comment by sun-peach** [comment]
@kamo-naoyuki Thank you for your reply, and sorry for the confusion.

My system intend to:
1. input speech into the encoder;
2. get the representation from decoder (the representation can be hidden values of one intermediate layer in the decoder, or it can be the text sequence output from the top layer of the decoder)
3. The representation is fed to several layers for further processing
4. output for classification (not ASR task).

The 1. and 2. are a ASR system pretrained with espnet. Now I want to load the pretrained block (encoder and decoder) into my system. I will add 3. and 4. on top of the pretrained block. 

My question is, how can I add further processing layers on top of the espnet model? should I do like this:
```
def forward(self, x):
    output_from_espnet_model = EspnetModel(x)
    final_output = nn.linear(4332, 3)(output_from_espnet_model)
    return final_output
```

**Comment by kamo-naoyuki** [comment]
If you don't intend to use the  transcription at the training, then some recognization system is required, but there are still some ambiguity. Which will you use beam-search or greedy-search?
Furthermore, do you need backward mode for the output from decoder? ESPnet2 doesn't provide it. Use espnet1.
If you don't need backward and you'll use beam-search, then refer https://github.com/espnet/espnet/blob/master/espnet2/bin/asr_inference.py

> (This is what I guess for output from the top layer of the espnet model. what about the output is from an intermediate layer?)

In general, in pytorch, to get the hidden vector from the intermediate layers, you always have to define a function for it by yourself. We can't provide functions for such specific situation.  
However, it's easy. What you have to do is just implementing a function to get the output from the decoder when recognition.

**Comment by sun-peach** [comment]
@kamo-naoyuki Thank you for the suggestion.

I don't use the text after beam search, but use the vector sequence from softmax layer (the characters distribution). I can use this as text embedding and send to further layer.

I need to do training, which means backward is needed. The error will be back-propagated from the further processing layers, but it is not token sequence. You can imaging I am adding a Transformer on top of the decoder. The transformer doing classification job on the utterance level. I would like to train the Transformer and the EspnetModel joint for the utterance classification.

I don't have the text sequence label and only have the classification label on the utterance level. In such a situation, do you have any further suggestion? Thank you once again.

**Comment by kamo-naoyuki** [comment]
If the transcription is not used, characters distribution doesn't exist. 
You may misunderstand what encoder-decoder model is.

**Comment by sun-peach** [comment]
@kamo-naoyuki Thank you.

When I pre-train them, everything is in the ASR training way, which means I have the transcription. Now I finish the training, and would like to incorporate the block into the system and retrain it. The top layer is not modeling the character distribution any more, but modeling the hidden values' distribution. Is this feasible to do?

Thank you very much!

**Comment by kamo-naoyuki** [comment]
You don't understand what I said.  

I say same things again.
ASR model requires transcription for training even if you just use the hidden vector with not ASR criterion. If you don't intend to use transcription, there are no explicit way. 

You should study about encoder-decoder model.

**Comment by sw005320** [comment]
Hi @sun-peach,

You could probably start to read some of our papers listed in https://github.com/espnet/espnet#references
(I'm actually working on (very slow though) writing a review paper about e2e ASR, but not yet ready...)

**Comment by sun-peach** [comment]
@kamo-naoyuki @sw005320 Thank you very much! I will go to read the materials.
